File: C#/WindowsPhone/LGame-XNA-lib/Loon.Utils/TextureUtils.cs
using Loon.Core.Graphics.Opengl;
using Loon.Core.Graphics;
using Loon.Core.Graphics.Device;
using Microsoft.Xna.Framework;
using Loon.Core;
using System;
namespace Loon.Utils
{
    public class TextureUtils
    {

        public static LTexture FilterColor(string res, LColor height)
        {
            return TextureUtils.FilterColor(res, height, Loon.Core.Graphics.Opengl.LTexture.Format.DEFAULT);
        }

        public static LTexture FilterColor(string res, LColor height, Loon.Core.Graphics.Opengl.LTexture.Format format)
        {
            uint color = height.GetRGB();
            LImage tmp = LImage.CreateImage(res);
            LImage image = LImage.CreateImage(tmp.GetWidth(), tmp.GetHeight(), true);
            LGraphics g = image.GetLGraphics();
            g.DrawImage(tmp, 0, 0);
            g.Dispose();
            if (tmp != null)
            {
                tmp.Dispose();
                tmp = null;
            }
            Color[] pixels = image.GetPixels();
            int size = pixels.Length;
            for (int i = 0; i < size; i++)
            {
                if (pixels[i].PackedValue == color)
                {
                    pixels[i].PackedValue = LSystem.TRANSPARENT;
                }
            }
            image.SetFormat(format);
            image.SetPixels(pixels, image.GetWidth(), image.GetHeight());
            LTexture texture = image.GetTexture();
            if (image != null)
            {
                image.Dispose();
                image = null;
            }
            return texture;
        }

        public static LTexture FilterLimitColor(string res, LColor start, LColor end)
        {
            return TextureUtils.FilterLimitColor(res, start, end, Loon.Core.Graphics.Opengl.LTexture.Format.DEFAULT);
        }

        public static LTexture FilterLimitColor(string res, LColor start,
                LColor end, Loon.Core.Graphics.Opengl.LTexture.Format format)
        {
            int sred = start.R;
            int sgreen = start.G;
            int sblue = start.B;
            int ered = end.R;
            int egreen = end.G;
            int eblue = end.B;
            LImage tmp = LImage.CreateImage(res);
            LImage image = LImage.CreateImage(tmp.GetWidth(), tmp.GetHeight(), true);
            LGraphics g = image.GetLGraphics();
            g.DrawImage(tmp, 0, 0);
            g.Dispose();
            if (tmp != null)
            {
                tmp.Dispose();
                tmp = null;
            }
            Color[] pixels = image.GetPixels();
            int size = pixels.Length;
            for (int i = 0; i < size; i++)
            {
                Color pixel = pixels[i];
                if ((pixel.R >= sred && pixel.G >= sgreen && pixel.B >= sblue)
                        && (pixel.R <= ered && pixel.G <= egreen && pixel.B <= eblue))
                {
                    pixels[i].PackedValue = LSystem.TRANSPARENT;
                }
            }
            image.SetFormat(format);
            image.SetPixels(pixels, image.GetWidth(), image.GetHeight());
            LTexture texture = image.GetTexture();
            if (image != null)
            {
                image.Dispose();
                image = null;
            }
            return texture;
        }

        public static LTexture FilterColor(string res, Color[] colors)
        {
            return TextureUtils.FilterColor(res, colors, Loon.Core.Graphics.Opengl.LTexture.Format.DEFAULT);
        }

        public static LTexture FilterColor(string res, Color[] colors, Loon.Core.Graphics.Opengl.LTexture.Format format)
        {
            LImage tmp = LImage.CreateImage(res);
            LImage image = LImage.CreateImage(tmp.GetWidth(), tmp.GetHeight(), true);
            LGraphics g = image.GetLGraphics();
            g.DrawImage(tmp, 0, 0);
            g.Dispose();
            if (tmp != null)
            {
                tmp.Dispose();
                tmp = null;
            }
            Color[] pixels = image.GetPixels();
            int size = pixels.Length;
            for (int i = 0; i < size; i++)
            {
                for (int j = 0; j < colors.Length; j++)
                {
                    if (pixels[i].Equals(colors[j]))
                    {
                        pixels[i].PackedValue = LSystem.TRANSPARENT;
                    }
                }
            }
            image.SetFormat(format);
            image.SetPixels(pixels, image.GetWidth(), image.GetHeight());
            LTexture texture = image.GetTexture();
            if (image != null)
            {
                image.Dispose();
                image = null;
            }
            return texture;
        }

        public static LTexture LoadTexture(string fileName)
        {
            return LTextures.LoadTexture(fileName);
        }

        public static LTexture[] GetSplitTextures(string fileName, int width,
                int height)
        {
            return GetSplitTextures(LTextures.LoadTexture(fileName), width, height);
        }

        public static LTexture[] GetSplitTextures(LTexture image, int width,
                int height)
        {
            if (image == null)
            {
                return null;
            }
            if (!image.IsLoaded())
            {
                image.LoadTexture();
            }
            int frame = 0;
            int wlength = image.GetWidth() / width;
            int hlength = image.GetHeight() / height;
            int total = wlength * hlength;
            LTexture[] images = new LTexture[total];
            for (int y = 0; y < hlength; y++)
            {
                for (int x = 0; x < wlength; x++)
                {
                    images[frame] = image.GetSubTexture((x * width), (y * height),
                            width, height);
                    frame++;
                }
            }
            return images;
        }

        public static LTexture[][] GetSplit2Textures(string fileName, int width,
                int height)
        {
            return GetSplit2Textures(LTextures.LoadTexture(fileName), width, height);
        }

        public static LTexture[][] GetSplit2Textures(LTexture image, int width,
                int height)
        {
            if (image == null)
            {
                return null;
            }
            if (!image.IsLoaded())
            {
                image.LoadTexture();
            }
            int wlength = image.GetWidth() / width;
            int hlength = image.GetHeight() / height;
            LTexture[][] textures = (LTexture[][])CollectionUtils.XNA_CreateJaggedArray(typeof(LTexture), wlength, hlength);
            for (int y = 0; y < hlength; y++)
            {
                for (int x = 0; x < wlength; x++)
                {
                    textures[x][y] = image.GetSubTexture((x * width), (y * height),
                            width, height);
                }
            }
            return textures;
        }

        public static LTexture[] GetDivide(string fileName, int count, int[] width,
                int[] height)
        {
            if (count <= 0)
            {
                throw new Exception("Divide");
            }
            LTexture image = LTextures.LoadTexture(fileName);
            if (image == null)
            {
                return null;
            }
            if (!image.IsLoaded())
            {
                image.LoadTexture();
            }

            if (width == null)
            {
                width = new int[count];
                int w = image.GetWidth();
                for (int j = 0; j < count; j++)
                {
                    width[j] = w / count;
                }
            }
            if (height == null)
            {
                height = new int[count];
                int h = image.GetHeight();
                for (int i = 0; i < count; i++)
                {
                    height[i] = h;
                }
            }
            LTexture[] images = new LTexture[count];
            int offsetX = 0;
            for (int i = 0; i < count; i++)
            {
                images[i] = image.GetSubTexture(offsetX, 0, width[i], height[i]);
                offsetX += width[i];
            }
            return images;
        }

        public static LTexture[] GetDivide(string fileName, int count)
        {
            return GetDivide(fileName, count, null, null);
        }

        public static LTexture CreateTexture(int width, int height, LColor c)
        {
            LImage image = LImage.CreateImage(width, height, false);
            LGraphics g = image.GetLGraphics();
            g.SetColor(c);
            g.FillRect(0, 0, width, height);
            g.Dispose();
            LTexture tex2d = image.GetTexture();
            if (image != null)
            {
                image.Dispose();
                image = null;
            }
            return tex2d;
        }

    }
}

File: Java/Android/OpenGLES/source/core/src/loon/core/graphics/opengl/TextureUtils.java
package loon.core.graphics.opengl;

import loon.core.LSystem;
import loon.core.graphics.LColor;
import loon.core.graphics.LImage;
import loon.core.graphics.device.LGraphics;
import loon.core.graphics.opengl.LTexture.Format;

/**
 * Copyright 2008 - 2012
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loon
 * @author cping
 * @email：javachenpeng@yahoo.com
 * @version 0.3.3
 */
public class TextureUtils {

	public static LTexture filterColor(String res, LColor height) {
		return TextureUtils.filterColor(res, height, Format.DEFAULT);
	}

	public static LTexture filterColor(String res, LColor height, Format format) {
		int color = height.getRGB();
		LImage tmp = new LImage(res);
		LImage image = new LImage(tmp.getWidth(), tmp.getHeight(), true);
		LGraphics g = image.getLGraphics();
		g.drawImage(tmp, 0, 0);
		g.dispose();
		if (tmp != null) {
			tmp.dispose();
			tmp = null;
		}
		int[] pixels = image.getPixels();
		int size = pixels.length;
		for (int i = 0; i < size; i++) {
			if (pixels[i] == color) {
				pixels[i] = 0xffffff;
			}
		}
		image.setFormat(format);
		image.setPixels(pixels, image.getWidth(), image.getHeight());
		LTexture texture = image.getTexture();
		if (image != null) {
			image.dispose();
			image = null;
		}
		return texture;
	}

	public static LTexture filterLimitColor(String res, LColor start, LColor end) {
		return TextureUtils.filterLimitColor(res, start, end, Format.DEFAULT);
	}

	public static LTexture filterLimitColor(String res, LColor start,
			LColor end, Format format) {
		int sred = start.getRed();
		int sgreen = start.getGreen();
		int sblue = start.getBlue();
		int ered = end.getRed();
		int egreen = end.getGreen();
		int eblue = end.getBlue();
		LImage tmp = new LImage(res);
		LImage image = new LImage(tmp.getWidth(), tmp.getHeight(), true);
		LGraphics g = image.getLGraphics();
		g.drawImage(tmp, 0, 0);
		g.dispose();
		if (tmp != null) {
			tmp.dispose();
			tmp = null;
		}
		int[] pixels = image.getPixels();
		int size = pixels.length;
		for (int i = 0; i < size; i++) {
			int[] rgbs = LColor.getRGBs(pixels[i]);
			if ((rgbs[0] >= sred && rgbs[1] >= sgreen && rgbs[2] >= sblue)
					&& (rgbs[0] <= ered && rgbs[1] <= egreen && rgbs[2] <= eblue)) {
				pixels[i] = 0xffffff;
			}
		}
		image.setFormat(format);
		image.setPixels(pixels, image.getWidth(), image.getHeight());
		LTexture texture = image.getTexture();
		if (image != null) {
			image.dispose();
			image = null;
		}
		return texture;
	}

	public static LTexture filterColor(String res, int[] colors) {
		return TextureUtils.filterColor(res, colors, Format.DEFAULT);
	}

	public static LTexture filterColor(String res, int[] colors, Format format) {
		LImage tmp = new LImage(res);
		LImage image = new LImage(tmp.getWidth(), tmp.getHeight(), true);
		LGraphics g = image.getLGraphics();
		g.drawImage(tmp, 0, 0);
		g.dispose();
		if (tmp != null) {
			tmp.dispose();
			tmp = null;
		}
		int[] pixels = image.getPixels();
		int size = pixels.length;
		for (int i = 0; i < size; i++) {
			for (int j = 0; j < colors.length; j++) {
				if (pixels[i] == colors[j]) {
					pixels[i] = 0xffffff;
				}
			}
		}
		image.setFormat(format);
		image.setPixels(pixels, image.getWidth(), image.getHeight());
		LTexture texture = image.getTexture();
		if (image != null) {
			image.dispose();
			image = null;
		}
		return texture;
	}

	public static LTexture loadTexture(String fileName) {
		return LTextures.loadTexture(fileName);
	}

	public static LTexture[] getSplitTextures(String fileName, int width,
			int height) {
		return getSplitTextures(LTextures.loadTexture(fileName), width, height);
	}

	public static LTexture[] getSplitTextures(LTexture image, int width,
			int height) {
		if (image == null) {
			return null;
		}

		if (LSystem.isThreadDrawing()) {
			image.loadTexture();
		}

		int frame = 0;
		int wlength = image.getWidth() / width;
		int hlength = image.getHeight() / height;
		int total = wlength * hlength;
		LTexture[] images = new LTexture[total];
		for (int y = 0; y < hlength; y++) {
			for (int x = 0; x < wlength; x++) {
				images[frame] = image.getSubTexture((x * width), (y * height),
						width, height);
				frame++;
			}
		}
		return images;
	}

	public static LTexture[][] getSplit2Textures(String fileName, int width,
			int height) {
		return getSplit2Textures(LTextures.loadTexture(fileName), width, height);
	}

	public static LTexture[][] getSplit2Textures(LTexture image, int width,
			int height) {
		if (image == null) {
			return null;
		}

		if (LSystem.isThreadDrawing()) {
			image.loadTexture();
		}

		int wlength = image.getWidth() / width;
		int hlength = image.getHeight() / height;
		LTexture[][] textures = new LTexture[wlength][hlength];
		for (int y = 0; y < hlength; y++) {
			for (int x = 0; x < wlength; x++) {
				textures[x][y] = image.getSubTexture((x * width), (y * height),
						width, height);
			}
		}
		return textures;
	}

	/**
	 * 0.3.2版起新增的分割图片方法，与上述近似作用的Split函数不同的是，可以指定个别图块大小。
	 * 
	 * @param fileName
	 * @param division
	 * @param width
	 * @param height
	 * @return
	 */
	public static LTexture[] getDivide(String fileName, int count, int[] width,
			int[] height) {
		if (count <= 0) {
			throw new IllegalArgumentException();
		}
		LTexture image = LTextures.loadTexture(fileName);
		if (image == null) {
			return null;
		}

		if (LSystem.isThreadDrawing()) {
			image.loadTexture();
		}

		if (width == null) {
			width = new int[count];
			int w = image.getWidth();
			for (int j = 0; j < count; j++) {
				width[j] = w / count;
			}
		}
		if (height == null) {
			height = new int[count];
			int h = image.getHeight();
			for (int i = 0; i < count; i++) {
				height[i] = h;
			}
		}
		LTexture[] images = new LTexture[count];
		int offsetX = 0;
		for (int i = 0; i < count; i++) {
			images[i] = image.getSubTexture(offsetX, 0, width[i], height[i]);
			offsetX += width[i];
		}
		return images;
	}

	/**
	 * 0.3.2版起新增的分割图片方法，成比例切分图片为指定数量
	 * 
	 * @param fileName
	 * @param count
	 * @return
	 */
	public static LTexture[] getDivide(String fileName, int count) {
		return getDivide(fileName, count, null, null);
	}

	/**
	 * 创建一张指定色彩的纹理
	 * 
	 * @param width
	 * @param height
	 * @param c
	 * @return
	 */
	public static LTexture createTexture(int width, int height, LColor c) {
		LImage image = new LImage(width, height, false);
		LGraphics g = image.getLGraphics();
		g.setColor(c);
		g.fillRect(0, 0, width, height);
		g.dispose();
		LTexture tex2d = image.getTexture();
		if (image != null) {
			image.dispose();
			image = null;
		}
		return tex2d;
	}

}
